### PR TITLE
Feature/schedule creation form

### PR DESCRIPTION
**Pull request by KalePitts**
schedule creation form pull request

### CODE REVIEW

File: Unpluq Extension Code/schedule-page/js/schedule_creation.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
+document.addEventListener('DOMContentLoaded', function () {
+    // Add click event listener to the 'Submit' label
+    var save_button = document.getElementById('schedule_save');
```
Comment (JornR94) [inline Unpluq Extension Code/schedule-page/js/schedule_creation.js:3]: For uniformity, please use camelCase instead of snake case (e.g. `saveButton` instead of `save_button` for the variable declaration. Also, Javascript style guide informs to use `let` or `const` (probably `const` in this case?), never var for local variables

File: Unpluq Extension Code/schedule-page/js/schedule_creation.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    save_button.addEventListener('click', function (event) {
+        // save_button is actually a link. the line below is to prevent it from navigating before it fetches input
```
Comment (JornR94) [inline Unpluq Extension Code/schedule-page/js/schedule_creation.js:6]: Great that you commented this, makes your code easy to follow!

File: Unpluq Extension Code/schedule-page/js/schedule_creation.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    else{
+        var startTimeHHmm = "00:00";
+        var endTimeHHmm = "23:59";
+    }
+    var checkedDays = fetchDays();
+    var blockedWebsites = fetchBlockedWebsites();
+    var barrierType = 4;
+    var barrierDifficulty = parseInt(document.getElementById('difficulty_range').value);
+    var newSchedule = [scheduleName, startTimeHHmm, endTimeHHmm, checkedDays, blockedWebsites, barrierType, barrierDifficulty];
```
Comment (JornR94) [inline Unpluq Extension Code/schedule-page/js/schedule_creation.js:35]: Let's create a Schedule class to use a schedule object instead of using an array. The biggest problems with this is that it is error prone since you have to remember the position of variables in the array and it becomes less readable (what is `schedule[0]`? `schedule.getName()` would be clearer). For writing to / reading from storage we can probably use (de)serializers to help

File: Unpluq Extension Code/schedule-page/js/schedule_creation.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    // test for 24-hour time format
+    const pattern = /^([01]?[0-9]|2[0-3]):[0-5][0-9]$/;
+    if (!pattern.test(newSchedule[1]) || !pattern.test(newSchedule[2])) {
+        callback(false);
+        return;
+    }
+
+    if(!newSchedule[3].includes(true)){
```
Comment (JornR94) [inline Unpluq Extension Code/schedule-page/js/schedule_creation.js:91]: The above two if statements were formatted correctly, but this one needs spacing

File: Unpluq Extension Code/schedule-page/js/schedule_creation.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+function checkValidInput(newSchedule, callback) {
+    if (newSchedule[0].trim() === "") {
+        callback(false);
+        return;
+    }
+
+    // test for 24-hour time format
+    const pattern = /^([01]?[0-9]|2[0-3]):[0-5][0-9]$/;
```
Comment (JornR94) [inline Unpluq Extension Code/schedule-page/js/schedule_creation.js:85]: Is this regex correct? Would be good to create some unit tests for these validation method, since it's easy to verify if they're working by checking the edge cases

File: Unpluq Extension Code/schedule-page/js/schedule_creation.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    if (newSchedule[4].length === 0) {
+        callback(false);
+        return;
+    }
+
+    chrome.storage.sync.get(['blockSchedules'], function (result) {
+        var existingblockSchedules = result.blockSchedules || [];
+
+        // Check if the first entry of the input array matches any in existingblockSchedules
+        var isDuplicate = existingblockSchedules.some(function (schedule) {
```
Comment (JornR94) [inline Unpluq Extension Code/schedule-page/js/schedule_creation.js:104]: existingblockSchedules -> existingBlockSchedules

File: Unpluq Extension Code/schedule-page/js/schedule_creation.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            return schedule[0].trim() === newSchedule[0].trim();
```
Comment (JornR94) [inline Unpluq Extension Code/schedule-page/js/schedule_creation.js:105]: It'd be good to already call .trim() on the schedule name when they're being stored, so that we know we always have a trimmed schedule name after saving a schedule

Comment (JornR94) [reply Unpluq Extension Code/schedule-page/js/schedule_creation.js:105]: Next to that (but I think you know this) it'd be good to check if the schedule exists before storing anything to storage

**Review by JornR94 — CHANGES_REQUESTED** [review]
Nice job Kaleb! You put in a lot of work and that definitely shows.
There are a couple of important things to fix and nip in the bud early on:

- Javascript styling: your styling is not very uniform - it's important to pay attention to those details since it improves readability a lot throughout the project. Please see https://google.github.io/styleguide/jsguide.html#features-local-variable-declarations
- Schedule storage: some comments about using a Schedule class instead of array and unit testing the verification methods

Important to fix this before moving on with other things, and to coordinate about the schedule storage with the others as well after these changes have been made (whether you or someone else on the team makes them).

Overall really good job. Thanks, and have a good winter break!
